sp/src/vscript/vscript_bindings_base.cpp

```diff
@@ -259,125 +264,132 @@ void KeyValues_SubKeysToTable( KeyValues *pKV, HSCRIPT hTable )
 	return hScriptInstance;
 }
 
 const char *CScriptKeyValues::ScriptGetName()
 {
-	const char *psz = m_pKeyValues->GetName();
+	const char *psz = GetKeyValues()->GetName();
 	return psz;
 }
 
 int CScriptKeyValues::ScriptGetInt()
 {
-	int i = m_pKeyValues->GetInt();
+	int i = GetKeyValues()->GetInt();
 	return i;
 }
 
 float CScriptKeyValues::ScriptGetFloat()
 {
-	float f = m_pKeyValues->GetFloat();
+	float f = GetKeyValues()->GetFloat();
 	return f;
 }
 
 const char *CScriptKeyValues::ScriptGetString()
 {
-	const char *psz = m_pKeyValues->GetString();
+	const char *psz = GetKeyValues()->GetString();
 	return psz;
 }
 
 bool CScriptKeyValues::ScriptGetBool()
 {
-	bool b = m_pKeyValues->GetBool();
+	bool b = GetKeyValues()->GetBool();
 	return b;
 }
 
 
 void CScriptKeyValues::ScriptSetKeyValueInt( const char *pszName, int iValue )
 {
-	m_pKeyValues->SetInt( pszName, iValue );
+	GetKeyValues()->SetInt( pszName, iValue );
 }
 
 void CScriptKeyValues::ScriptSetKeyValueFloat( const char *pszName, float flValue )
 {
-	m_pKeyValues->SetFloat( pszName, flValue );
+	GetKeyValues()->SetFloat( pszName, flValue );
 }
 
 void CScriptKeyValues::ScriptSetKeyValueString( const char *pszName, const char *pszValue )
 {
-	m_pKeyValues->SetString( pszName, pszValue );
+	GetKeyValues()->SetString( pszName, pszValue );
 }
 
 void CScriptKeyValues::ScriptSetKeyValueBool( const char *pszName, bool bValue )
 {
-	m_pKeyValues->SetBool( pszName, bValue );
+	GetKeyValues()->SetBool( pszName, bValue );
 }
 
 void CScriptKeyValues::ScriptSetName( const char *pszValue )
 {
-	m_pKeyValues->SetName( pszValue );
+	GetKeyValues()->SetName( pszValue );
 }
 
 void CScriptKeyValues::ScriptSetInt( int iValue )
 {
-	m_pKeyValues->SetInt( NULL, iValue );
+	GetKeyValues()->SetInt( NULL, iValue );
 }
 
 void CScriptKeyValues::ScriptSetFloat( float flValue )
 {
-	m_pKeyValues->SetFloat( NULL, flValue );
+	GetKeyValues()->SetFloat( NULL, flValue );
 }
 
 void CScriptKeyValues::ScriptSetString( const char *pszValue )
 {
-	m_pKeyValues->SetString( NULL, pszValue );
+	GetKeyValues()->SetString( NULL, pszValue );
 }
 
 void CScriptKeyValues::ScriptSetBool( bool bValue )
 {
-	m_pKeyValues->SetBool( NULL, bValue );
+	GetKeyValues()->SetBool( NULL, bValue );
 }
 
 
 // constructors
-CScriptKeyValues::CScriptKeyValues( KeyValues *pKeyValues = NULL )
+CScriptKeyValues::CScriptKeyValues( KeyValues *pKeyValues = NULL, bool bBorrow = false ) :
+	m_pBase( NULL )
 {
 	if (pKeyValues == NULL)
 	{
-		m_pKeyValues = new KeyValues("CScriptKeyValues");
-	}
-	else
-	{
-		m_pKeyValues = pKeyValues;
+		pKeyValues = new KeyValues("CScriptKeyValues");
+		// Borrowed new memory doesn't make sense, are you trying to leak?
+		Assert( !bBorrow );
 	}
+
+	m_pSelf = new KeyValues_RC( pKeyValues, bBorrow );
 }
 
 // destructor
 CScriptKeyValues::~CScriptKeyValues( )
 {
-	if (m_pKeyValues)
+	// Children are always borrowed
+	Assert( !m_pBase || m_pSelf->borrow );
```

Comment:
Not the case for registering C++ managed KeyValues, such as in `CLogicExternalData::ScriptGetKeyValues()`. Though in this particular case the entity can die while `CScriptKeyValues` lives. Registering a copy would make editing the result of `GetKeyValues()` not change `m_pRoot`, but require `SetKeyValues()` to update. A cumbersome workaround would be `CLogicExternalData` keeping copies of script instances it registers. I think @Blixibon could give his input on this since it was his design choice.

Logically it makes sense, and it works given the borrowed memory isn't independently freed before the script VM.